src/training/graph_group.h

```diff
@@ -58,19 +65,32 @@ class AsyncGraphGroup : public GraphGroup {
         return;
 
       // @TODO read guard on parameters
-      std::lock_guard<std::mutex> guard(sync_);
-      oldParams->copyFrom(params_);
+      int pos = 0;
+      for (int idx = 0; idx < devices_.size(); idx++) {
+        std::lock_guard<std::mutex> guard( shardSync_[idx] );
+        oldParams->subtensor(pos , params_[idx]->size())->copyFrom(params_[idx]);
+        pos += shardSize_;
```

Comment:
Ignore this for a moment. I am testing different solutions, so far all of them have bad convergence behavior.